packages/flutter_tools/lib/src/android/android_sdk.dart

```diff
@@ -417,36 +470,64 @@ class AndroidSdk {
     required Platform platform,
   }) {
     if (androidStudio?.javaPath != null) {
+      globals.printTrace("Using Android Studio's java.");
       return fileSystem.path.join(androidStudio!.javaPath!, 'bin', 'java');
     }
 
-    final String? javaHomeEnv = platform.environment[_javaHomeEnvironmentVariable];
+    final String? javaHomeEnv =
+        platform.environment[_javaHomeEnvironmentVariable];
     if (javaHomeEnv != null) {
       // Trust JAVA_HOME.
+      globals.printTrace('Using JAVA_HOME.');
       return fileSystem.path.join(javaHomeEnv, 'bin', 'java');
     }
 
     // MacOS specific logic to avoid popping up a dialog window.
     // See: http://stackoverflow.com/questions/14292698/how-do-i-check-if-the-java-jdk-is-installed-on-mac.
     if (platform.isMacOS) {
       try {
-        final String javaHomeOutput = globals.processUtils.runSync(
-          <String>['/usr/libexec/java_home', '-v', '1.8'],
-          throwOnError: true,
-          hideStdout: true,
-        ).stdout.trim();
+        // -v Filter versions (as if JAVA_VERSION had been set in the environment).
+        // It is unlikley that filtering to java version 1.8 is the right
+        // decision here. That said, trying this on a mac shows the same jdk
+        // path no matter what input is passed.
+        final String javaHomeOutput = globals.processUtils
+            .runSync(
+              <String>['/usr/libexec/java_home', '-v', '1.8'],
+              throwOnError: true,
+              hideStdout: true,
+            )
+            .stdout
+            .trim();
         if (javaHomeOutput.isNotEmpty) {
           final String javaHome = javaHomeOutput.split('\n').last.trim();
+          globals.printTrace('Using mac JAVA_HOME.');
           return fileSystem.path.join(javaHome, 'bin', 'java');
         }
-      } on Exception { /* ignore */ }
+      } on Exception {/* ignore */}
     }
 
     // Fallback to PATH based lookup.
-    return operatingSystemUtils.which(_javaExecutable)?.path;
+    final String? pathJava = operatingSystemUtils.which(_javaExecutable)?.path;
+    if (pathJava != null) {
+      globals.printTrace('Using path java.');
```

Comment:
tiny nit ("path java" took me a few seconds to parse&understand than "java from PATH")

```suggestion
      globals.printTrace('Using java from PATH.');
```